orocos_kdl/tests/solvertest.cpp

```diff
@@ -1550,3 +1558,167 @@ void SolverTest::FdAndVereshchaginSolversConsistencyTest()
 
     return;
 }
+
+void SolverTest::ExternalWrenchEstimatorTest()
+{
+    /**
+     * Closed-loop test for the external wrench estimator class:
+     * Simple controlled behaviour of the robot subjected to an external force is simulated.
+     * The external wrench estimator is called in each iteration of the control loop 
+     * so to converge to final wrench value.
+     * In the end, estimated wrench is compared to the ground-truth values of the simulated wrench. 
+     */
+    std::cout << "KDL External Wrench Estimator Test" << std::endl;
+
+    int ret;
+    double eps = 0.85;
```

Comment:
Well, the first thing here is that this EPS have a slightly different purpose than the EPSes in tests of other solvers. Meaning that those other EPSes have the purpose for taking care of the differences that came to exist due to e.g. floating imprecisions, different compilers (or same compiler but different flags) used in building the code between different machines (OS), etc. This particular EPS of 0.85 is there to cover those imperfections as well but, more importantly, it's there to take into account the noise in estimated signals (the differences between estimated and ground-truth wrenches) caused by other computations in this test method (ones coming from the implemented PID controller and dynamics simulator), not just those coming from the estimator itself.

More specifically, these state-estimation techniques are, in general, more closely related to the family of controllers (such as PID, ABAG, Sliding Mode control - cf. previously mentioned reference [2], etc.) then to the dynamics solvers... even though, both, estimators and dynamics algorithms, are deterministic in their nature (at least those types used in KDL). The reason is that the estimators act on a set of multiple (different time-step related) input signals to converge to a valid output. Which is in contrast to dynamics algorithms (being analytical or numerical) which act on a single input signal to converge to a valid output. Meaning that with each new iteration, the estimator signal is affected by a new noise value. In reality, the noise in input signals can originate from different sources (e.g. unmodelled or not-correctly modelled dynamics parameters of the robot, friction is the most common example... low-quality joint encoders, etc.). However, in this particular case (in implemented test function) the noise mainly originates from:
* the implemented dynamics simulator. Due to the simple integrator implemented (Symplectic Euler), the simulated joint state drifts away quite quickly.
* the implemented PID controller. Simple computed-torque control in combination with low-time-effort-tuned gains introduces additional "fluctuations" in robot's behaviour.
* a pseudo-inverse of the jacobian matrix used to propagate estimated joint torques to Cart. wrenches introduce deviations by itself. This is something that is governed by nature (math-physics)... and it does not leave us much room to change. Plus, Jacobians act as noise amplifiers as well.

In order to improve the test, such that the EPS can be selected smaller, the first two points mentioned above, i.e. the control-simulation loop can be improved. E.g. instead of using Symplectic Euler, Runge-Kutta 4th order integrator can be added and/or instead of going with simple computed-torque joint-space control, more advanced Cartesian-control can be implemented, the control gains can be tuned better, etc. 

There might be other ways to improve this loop as well, but IMO, outside estimator's computations. In general, a user should expect that in a real-world setting the difference between estimated and ground-truth values will be larger than 0.85. In a real-world setting, there will be more noise sources (as previously mentioned, e.g. unmodelled or not-correctly modelled dynamics parameters of the robot, friction is the most common example... noisy joint encoders... the existence of real joint torque sensors, etc).

Reply:
Ok, so we abusing the EPS here a little bit. Therefore it is good to add a comment in the code.

Reply:
I have added a comment in code for this EPS.